src/tcache.c

```diff
@@ -121,6 +121,80 @@ tcache_gc_dalloc_postponed_event_wait(tsd_t *tsd) {
 	return TE_MIN_START_WAIT;
 }
 
+static inline void
+tcache_nfill_small_ctl_init(tcache_slow_t *tcache_slow, szind_t szind) {
+	cache_bin_fill_ctl_t *ctl =
+	    &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
+	ctl->base = 1;
+	ctl->offset = 0;
+}
+
+/*
+ * The number of items to be filled at a time for a given small bin is
+ * calculated by (ncached_max >> lg_fill_div).
+ * The actual ctl struct consists of two fields, i.e. base and offset,
+ * and the difference between the two(base - offset) is the final lg_fill_div.
+ * The base is adjusted during GC based on the traffic within a period of time,
+ * while the offset is updated in real time to handle the immediate traffic.
+ */
+static inline uint8_t
+tcache_nfill_small_lg_div_get(tcache_slow_t *tcache_slow, szind_t szind) {
+	cache_bin_fill_ctl_t *ctl =
+	    &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
+	assert(ctl->base > ctl->offset);
+	return (ctl->base - (opt_experimental_tcache_gc ? ctl->offset : 0));
+}
+
+static inline void
+tcache_nfill_small_gc_adjust(tcache_slow_t *tcache_slow, szind_t szind,
+    cache_bin_sz_t limit) {
+	cache_bin_fill_ctl_t *ctl =
+	    &tcache_slow->bin_fill_ctl_do_not_access_directly[szind];
+	assert(ctl->base > ctl->offset);
+	/* Reset the offset for the next GC period. */
+	ctl->offset = 0;
+	/*
+	 * limit == 0: indicates that the fill count should be increased,
+	 * i.e. lg_div(base) should be decreased.
+	 *
+	 * limit != 0: set to ncached_max, and indicates that the fill
+	 * count should be decreased, i.e. lg_div(base) should be increased.
+	 */
+	if (!limit && ctl->base > 1) {
+		/*
+		 * Increase fill count by 2X for small bins.  Make sure
+		 * lg_fill_div stays greater than 1.
+		 */
+		ctl->base--;
+	} else if (limit && ((limit >> ctl->base) > 1)) {
+		/*
+		 * Reduce fill count by 2X.  Limit lg_fill_div such that
+		 * the fill count is always at least 1.
+		 */
+		ctl->base++;
+	}
+}
+
+/* Update the offset, which is used to handle burst loads. */
+static inline void
+tcache_nfill_small_burst_adjust(tcache_slow_t *tcache_slow, szind_t szind,
+    bool may_have_burst) {
```

Comment:
Let's split this into the two functions? e.g. tcache_nfill_small_burst_reset and tcache_nfill_small_burst_trigger (or _prepare)?

We are not sharing much logic in the current adjust function, and it's a bit confusing since we have another _adjust function right above it.